### PR TITLE
Add more ARM32 instructions semantics

### PR DESCRIPTION
**Pull request by black-binary**
I added some instructions for arm32, most of which were modified from aarch32. But it is not fully tested. Unit tests will be added later on.

Would you please review the code? I'm not sure if I implemented the condition suffix in the correct way.  :)

### CODE REVIEW

**Comment by SweetVishnya** [comment]
We definitely should have some tests.

**Comment by black-binary** [comment]
@SweetVishnya  Sure, it's still working in progress now. And I'm looking into how to add unit tests. Seems like I have to write a script to generate python code for the unit tests.

**Comment by JonathanSalwan** [comment]
Thanks @black-binary for this MR! Let me few weeks to review this :). Adding as much as possible unittests to test corners cases will help for sure :).

Cheers!

**Comment by black-binary** [comment]
@JonathanSalwan I found a lot of bugs after adding unit tests for these instructions XD. Need some time to fix them.

**Comment by JonathanSalwan** [comment]
That's the power of unit tests :D! Don't forget to also check the thumb mode :). Ping me when you think the MR can be reviewed :)

**Comment by black-binary** [comment]
I can not write a unit test for UDIV/SDIV, because the Keystone failed to assemble UDIV/SDIV opcodes (KS_ERR_ASM_MISSINGFEATURE) so I can't generate the machine code.

Any good idea?

**Comment by SweetVishnya** [comment]
You can assemble them with other assembler manually and just insert raw bytes.

**Comment by black-binary** [comment]
@JonathanSalwan 

Most of the code has been tested, and I'm currently using these new features in my project(and they work well so far). Could you please review the code? Thanks :)

**Comment by JonathanSalwan** [comment]
Got it. Thanks for this MR! Let me few weeks to review and test the code.

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 1]
```diff
@@ -1360,6 +1449,42 @@ namespace triton {
...
+                          this->astCtxt->equal(
+                            op1,
+                            this->astCtxt->bv(0, op1->getBitvectorSize())
+                          ),
+                          op3,
+                          op2
+                        );
+
+          /* Create symbolic expression */
+          auto expr = this->symbolicEngine->createSymbolicExpression(inst, pcNode, dst, "CBZ operation - Program Counter");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:1473]: CBNZ*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 2]
```diff
@@ -1424,6 +1549,42 @@ namespace triton {
...
+          /* Create symbolic operands */
+          auto op1 = this->getArm32SourceOperandAst(inst, src1);
+          auto op2 = this->getArm32SourceOperandAst(inst, src2);
+
+          /* Create the semantics */
+          auto cond = this->getCodeConditionAst(inst);
+          auto node1 = this->astCtxt->bvadd(op1, op2);
+
+          /* Create symbolic expression */
+          auto expr = this->symbolicEngine->createSymbolicVolatileExpression(inst, node1, "CMP operation");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:1565]: CMN*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 3]
```diff
@@ -1814,76 +1975,95 @@ namespace triton {
...
-          auto cond  = this->getCodeConditionAst(inst);
-          auto node1 = this->astCtxt->ite(cond, op1, op2);
-          auto node2 = this->astCtxt->ite(cond, op3, op4);
+          auto node1 = this->astCtxt->zx(dst.getBitSize() - src.getBitSize(), op);
+          auto node2 = this->buildConditionalSemantics(inst, dst, node1);
 
           /* Create symbolic expression */
-          auto expr1 = this->symbolicEngine->createSymbolicExpression(inst, node1, dst1, "LDRD operation - LOAD access");
-          auto expr2 = this->symbolicEngine->createSymbolicExpression(inst, node2, dst2, "LDRD operation - LOAD access");
+          auto expr = this->symbolicEngine->createSymbolicExpression(inst, node2, dst, "LDRB operation - LOAD access");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:1993]: LDRH*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 4]
```diff
@@ -1814,76 +1975,95 @@ namespace triton {
...
+              if (imm.isSubtracted() == true) {
+                thenNode = this->astCtxt->bvsub(baseNode, this->astCtxt->sx(base.getBitSize() - imm.getBitSize(), immNode));
+              }
 
-            /* Spread taint */
-            this->spreadTaint(inst, cond, expr2, base, this->taintEngine->isTainted(base));
+              auto node2 = this->astCtxt->ite(cond, thenNode, baseNode);
+
+              /* Create symbolic expression */
+              auto expr2 = this->symbolicEngine->createSymbolicExpression(inst, node2, base, "LDRB operation - Post-indexed base register computation");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:2025]: LDRH*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 5]
```diff
@@ -1814,76 +1975,95 @@ namespace triton {
...
+              auto thenNode = this->astCtxt->bvadd(baseNode, regNode);
+
+              if (reg.isSubtracted() == true) {
+                thenNode = this->astCtxt->bvsub(baseNode, regNode);
+              }
+
+              auto node2 = this->astCtxt->ite(cond, thenNode, baseNode);
+
+              /* Create symbolic expression */
+              auto expr2 = this->symbolicEngine->createSymbolicExpression(inst, node2, base, "LDRB operation - Post-indexed base register computation");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:2047]: LDRH*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 6]
```diff
@@ -1814,76 +1975,95 @@ namespace triton {
...
             /* Create symbolic operands of the post computation */
             auto baseNode = this->symbolicEngine->getOperandAst(inst, base);
 
             /* Create the semantics of the base register */
-            auto node3 = this->astCtxt->ite(cond, inst.operands[2].getMemory().getLeaAst(), baseNode);
+            auto node3 = this->astCtxt->ite(cond, src.getMemory().getLeaAst(), baseNode);
 
             /* Create symbolic expression */
-            auto expr3 = this->symbolicEngine->createSymbolicExpression(inst, node3, base, "LDRD operation - Pre-indexed base register computation");
+            auto expr3 = this->symbolicEngine->createSymbolicExpression(inst, node3, base, "LDRB operation - Pre-indexed base register computation");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:2066]: LDRH*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 7]
```diff
@@ -3318,6 +4012,54 @@ namespace triton {
...
+        void Arm32Semantics::sxtb_s(triton::arch::Instruction& inst) {
+          auto& dst = inst.operands[0];
+          auto& src = inst.operands[1];
+
+          /* Create symbolic operands */
+          auto op = this->symbolicEngine->getOperandAst(inst, src);
+
+          /* Create the semantics */
+          auto node1 = this->astCtxt->sx(dst.getBitSize() - 8, this->astCtxt->extract(7, 0, op));
+          auto node2 = this->buildConditionalSemantics(inst, dst, node1);
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:4024]: Please remove space at the end of this line

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 8]
```diff
@@ -3318,6 +4012,54 @@ namespace triton {
...
+        void Arm32Semantics::sxth_s(triton::arch::Instruction& inst) {
+          auto& dst = inst.operands[0];
+          auto& src = inst.operands[1];
+
+          /* Create symbolic operands */
+          auto op = this->symbolicEngine->getOperandAst(inst, src);
+
+          /* Create the semantics */
+          auto node1 = this->astCtxt->sx(dst.getBitSize() - 16, this->astCtxt->extract(15, 0, op));
+          auto node2 = this->buildConditionalSemantics(inst, dst, node1);
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:4048]: Please remove the missing space at the end of this line

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 9]
```diff
@@ -3353,6 +4095,203 @@ namespace triton {
...
+                         this->astCtxt->zx(triton::bitsize::qword, op1),
+                         this->astCtxt->zx(triton::bitsize::qword, op2)
+                       );
+          auto lower = this->astCtxt->extract(triton::bitsize::dword-1, 0, mul);
+          auto upper = this->astCtxt->extract(triton::bitsize::qword-1, triton::bitsize::dword, mul);
+          auto node1 = this->astCtxt->ite(cond, lower, this->symbolicEngine->getOperandAst(inst, dst1));
+          auto node2 = this->astCtxt->ite(cond, upper, this->symbolicEngine->getOperandAst(inst, dst2));
+
+          /* Create symbolic expression */
+          auto expr1 = this->symbolicEngine->createSymbolicExpression(inst, node1, dst1, "SMULL(S) operation - Lower 32 bits of the result.");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:4215]: UMLL*

File: src/libtriton/arch/arm/arm32/arm32Semantics.cpp [hunk 10]
```diff
@@ -3353,6 +4095,203 @@ namespace triton {
...
+          auto expr2 = this->symbolicEngine->createSymbolicExpression(inst, node2, dst2, "SMULL(S) operation - Upper 32 bits of the result.");
```
Comment (JonathanSalwan) [inline src/libtriton/arch/arm/arm32/arm32Semantics.cpp:4216]: UMLL*

**Comment by JonathanSalwan** [comment]
Also can you verify that between each function there are 2 end of line. Otherwise the MR looks very good! Thanks

**Comment by black-binary** [comment]
Done :)

**Pull request merged**